plugins/inputs/modbus/modbus_test.go

```diff
@@ -1927,6 +1927,962 @@ func TestRequestsStartingWithOmits(t *testing.T) {
+			inputs: []rangeDefinition{
+				{0, 1, 1, 1, "INT16", false},
+				{1, 17, 1, 1, "INT16", true},
+				{18, 3, 1, 1, "INT16", false},
+				{21, maxsize - 23, 1, 1, "INT16", true},
+				{maxsize - 2, 5, 1, 1, "INT16", false},
+				{maxsize + 3, 39, 1, 1, "INT16", true},
+				{maxsize + 42, 2, 1, 1, "INT16", false},
+			},
+			expected: []requestExpectation{
+				{
+					fields: []rangeDefinition{
+						{start: 0, count: 1, length: 1},
+						{start: 18, count: 3, length: 1},
+					},
+					req: request{address: 0, length: 21},
+				},
+				{
+					fields: []rangeDefinition{
+						{start: maxsize - 2, count: 5, length: 1},
+						{start: maxsize + 42, count: 2, length: 1},
+					},
+					req: request{address: maxsize - 2, length: 46},
+				},
+			},
+		},
+		{
+			name: "large gaps filled with offset",
+			inputs: []rangeDefinition{
+				{18, 3, 1, 1, "INT16", false},
+				{21, maxsize - 23, 1, 1, "INT16", true},
+				{maxsize - 2, 5, 1, 1, "INT16", false},
+				{maxsize + 3, 39, 1, 1, "INT16", true},
+				{maxsize + 42, 2, 1, 1, "INT16", false},
+			},
+			expected: []requestExpectation{
+				{
+					fields: []rangeDefinition{{start: 18, count: 3, length: 1}},
+					req:    request{address: 18, length: 3},
+				},
+				{
+					fields: []rangeDefinition{
+						{start: maxsize - 2, count: 5, length: 1},
+						{start: maxsize + 42, count: 2, length: 1},
+					},
+					req: request{address: maxsize - 2, length: 46},
+				},
+			},
+		},
+		{
+			name: "from PR #11106",
+			inputs: []rangeDefinition{
+				{0, 2, 1, 1, "INT16", true},
+				{2, 1, 1, 1, "INT16", false},
+				{3, 2*maxsize + 1, 1, 1, "INT16", true},
+				{2*maxsize + 1, 1, 1, 1, "INT16", false},
+			},
+			expected: []requestExpectation{
+				{
+					fields: []rangeDefinition{{start: 2, count: 1, length: 1}},
+					req:    request{address: 2, length: 1},
+				},
+				{
+					fields: []rangeDefinition{{start: 2*maxsize + 1, count: 1, length: 1}},
+					req:    request{address: 2*maxsize + 1, length: 1},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Generate the input structure and the expectation
+			requestFields := generateRequestDefinitions(tt.inputs)
+			expected := generateExpectation(tt.expected)
+
+			// Setup the plugin
+			slaveID := byte(1)
+			plugin := Modbus{
+				Name:              "Test",
+				Controller:        "tcp://localhost:1502",
+				ConfigurationType: "request",
+				Log:               testutil.Logger{},
+			}
+			plugin.Requests = []requestDefinition{
+				{
+					SlaveID:      slaveID,
+					ByteOrder:    "ABCD",
+					RegisterType: "holding",
+					Optimization: "aggressive",
+					Fields:       requestFields,
+				},
+			}
+			require.NoError(t, plugin.Init())
+			require.NotEmpty(t, plugin.requests)
+			require.Contains(t, plugin.requests, slaveID)
+			requireEqualRequests(t, expected, plugin.requests[slaveID].holding)
+		})
+	}
+}
+
+func TestRequestOptimizationMaxInsertLarge(t *testing.T) {
```

Comment:
Please also add a small test for checking that invalid values of `MaxExtraRegisters` trigger an error. Use a separate test function for this!